python/pyspark/sql/functions.py

```diff
@@ -222,6 +222,45 @@ def sum_distinct(col):
     return _invoke_function_over_column("sum_distinct", col)
 
 
+def product(col, scale=1.0):
+    """
+    Aggregate function: returns the product of the values in a group.
+
+    .. versionadded:: 3.2.0
+
+    Parameters
+    ----------
+    col : str, :class:`Column`
+        column containing values to be multiplied together
+    scale : float
```

Comment:
I'm just looking at this again - why is this scale factor needed?

Reply:
Thanks, Sean, for picking this up again.

The scale-factor is there to allow the user some control over products that are likely to cause overflow. For example, if the user knows that the numbers in the product are typically about 100, they could set the scale to be 0.01. This would reduce the risk of the product overflowing when large numbers of terms are multiplied together. Obviously, the user would then have to take account of a factor of (0.01^N) in their subsequent calculations, but that should be far easier than dealing with overflow. By default, the scale is set to one, and the implementation treats this as a special case which doesn't involve additional multiplications by unity.

Reply:
I get that a very large float loses some precision as it only has so many significant digits to play with. But does scaling them make it more precise? each multiplication by 0.01 or whatever also introduces some error. It won't stop the final result from overflowing a double if in fact it's so large right? or is the idea that you don't want the actual product, but the scaled product? In that case can you compute `product(0.01 * $"col")` anyway, that kind of thing?

Reply:
Agreed, the `scale` parameter isn't there to improve precision. (Perhaps my "0.01" example wasn't ideal, 1/128 might have been a better choice that is less likely to cause noise in the least significant bits.) The aim, as you say, is to allow the user to compute the *scaled* product, on the assumption that they can allow for the *overall* scaling (i.e. the 0.01^N or 2^(-7N) in some other way at subsequent stages of their calculations. Clearly, there's no point in having this scaling within `product()` if the user just multiplies the result by (scale ^ -N).

Again, you're right that the user could just use `product(0.01 * $"col")`, but it seemed worth offering an overload that invites the user to consider using a scaling when they can predict something about the order-of-magnitude of their product, and might make their intention a little clearer than just multiplying `$"col"` by 0.01. This also allows a little bit more optimization if `scale` is the result of a calculation or a configuration parameter, and where it could turn out to be 1.

Reply:
I don't know it adds another arg that seems pretty niche, when you can express this the natural way with existing syntax.

Reply:
Not sure if you're comment is specifically about the Python API, or applies elsewhere too. It's very easy to simplify the Python & R APIs to remove the `scale` parameter, but removing that functionality from the underlying Scala code would be more involved. If possible, I'd like to keep this functionality available in Scala for those "niche" use-cases unless you feel strongly that they shouldn't be there.

Reply:
Yes, for all of the APIs. I mean, don't you just remove the params and simplify to the case where scale = 1?
@HyukjinKwon do you have an opinion on this one

Reply:
I would also avoid adding `scale`. `product(0.01 * $"col")` looks easy and straightforward enough.

Reply:
I've now removed the "scale" parameter from the `functions` API.